File: src/main/java/net/gegy1000/terrarium/server/world/generator/EarthChunkGenerator.java
package net.gegy1000.terrarium.server.world.generator;

import net.gegy1000.terrarium.server.capability.TerrariumCapabilities;
import net.gegy1000.terrarium.server.capability.TerrariumWorldData;
import net.gegy1000.terrarium.server.map.GenerationRegionHandler;
import net.gegy1000.terrarium.server.map.LatitudinalZone;
import net.gegy1000.terrarium.server.map.cover.CoverGenerator;
import net.gegy1000.terrarium.server.map.cover.CoverType;
import net.gegy1000.terrarium.server.map.cover.generator.primer.GlobChunkPrimer;
import net.gegy1000.terrarium.server.map.cover.generator.primer.GlobPrimer;
import net.gegy1000.terrarium.server.map.system.chunk.CoverChunkDataProvider;
import net.gegy1000.terrarium.server.map.system.chunk.HeightChunkDataProvider;
import net.gegy1000.terrarium.server.map.system.component.TerrariumComponentTypes;
import net.gegy1000.terrarium.server.util.ArrayUtils;
import net.gegy1000.terrarium.server.util.Coordinate;
import net.gegy1000.terrarium.server.util.Lazy;
import net.gegy1000.terrarium.server.world.EarthGenerationSettings;
import net.minecraft.block.BlockFalling;
import net.minecraft.block.state.IBlockState;
import net.minecraft.entity.EnumCreatureType;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.WorldEntitySpawner;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.chunk.Chunk;
import net.minecraft.world.chunk.ChunkPrimer;
import net.minecraft.world.gen.NoiseGeneratorPerlin;
import net.minecraftforge.event.ForgeEventFactory;
import net.minecraftforge.event.terraingen.PopulateChunkEvent;
import net.minecraftforge.event.terraingen.TerrainGen;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

public class EarthChunkGenerator extends CoveredChunkGenerator {
    public static final long ZONE_SCATTER_SEED = 3111632783370753319L;

    private final EarthGenerationSettings settings;
    private final boolean fastGenerate;

    private final World world;
    private final Random random;
    private final NoiseGeneratorPerlin coverNoise;
    private final PseudoRandomMap zoneScatterMap;

    private double[] depthBuffer = new double[256];
    private final IBlockState[] coverBlockBuffer = ArrayUtils.defaulted(new IBlockState[256], STONE);
    private final IBlockState[] fillerBlockBuffer = ArrayUtils.defaulted(new IBlockState[256], STONE);

    private Biome[] biomeBuffer = null;

    private final Lazy<GenerationRegionHandler> regionHandler;

    private final HeightChunkDataProvider heightProvider;
    private final CoverChunkDataProvider coverProvider;

    private final Map<CoverType, CoverGenerator> generators;

    public EarthChunkGenerator(World world, long seed, EarthGenerationSettings settings, boolean fastGenerate) {
        super(world, settings.getOceanHeight());
        this.world = world;

        this.settings = settings;
        this.fastGenerate = fastGenerate;

        this.random = new Random(seed);
        this.coverNoise = new NoiseGeneratorPerlin(this.random, 4);
        this.zoneScatterMap = new PseudoRandomMap(world, ZONE_SCATTER_SEED);

        this.heightProvider = new HeightChunkDataProvider(this.settings, TerrariumComponentTypes.HEIGHT);
        this.coverProvider = new CoverChunkDataProvider(this.settings, world, TerrariumComponentTypes.COVER);

        this.regionHandler = new Lazy<>(() -> {
            TerrariumWorldData capability = this.world.getCapability(TerrariumCapabilities.worldDataCapability, null);
            if (capability != null) {
                return capability.getRegionHandler();
            }
            throw new RuntimeException("Tried to load " + GenerationRegionHandler.class.getSimpleName() + " before it was present");
        });

        CoverType[] coverBuffer = this.coverProvider.getResultStore().getCoverData();
        int[] heightBuffer = this.heightProvider.getResultStore();
        this.generators = super.createGenerators(coverBuffer, heightBuffer, this.coverBlockBuffer, this.fillerBlockBuffer, false);
    }

    @Override
    public Chunk generateChunk(int chunkX, int chunkZ) {
        ChunkPrimer primer = this.generatePrimer(chunkX, chunkZ);
        this.biomeBuffer = this.populateBiomes(this.biomeBuffer, chunkX, chunkZ);

        Chunk chunk = new Chunk(this.world, primer, chunkX, chunkZ);

        if (this.biomeBuffer != null) {
            byte[] biomeArray = chunk.getBiomeArray();
            for (int i = 0; i < this.biomeBuffer.length; i++) {
                biomeArray[i] = (byte) Biome.getIdForBiome(this.biomeBuffer[i]);
            }
        }

        chunk.generateSkylightMap();

        return chunk;
    }

    public ChunkPrimer generatePrimer(int chunkX, int chunkZ) {
        this.random.setSeed(chunkX * 341873128712L + chunkZ * 132897987541L);

        ChunkPrimer primer = new ChunkPrimer();
        this.populateBlocks(primer, chunkX, chunkZ);
        this.generateBiome(primer, chunkX, chunkZ);

        return primer;
    }

    private void populateBlocks(ChunkPrimer primer, int chunkX, int chunkZ) {
        int oceanHeight = this.settings.getOceanHeight();
        this.heightProvider.populate(this.regionHandler.get(), this.world, chunkX << 4, chunkZ << 4);

        int[] heightData = this.heightProvider.getResultStore();
        for (int z = 0; z < 16; z++) {
            for (int x = 0; x < 16; x++) {
                int height = heightData[x + z * 16];
                for (int y = 1; y <= height; y++) {
                    primer.setBlockState(x, y, z, STONE);
                }
                if (height < oceanHeight) {
                    for (int y = height + 1; y <= oceanHeight; y++) {
                        primer.setBlockState(x, y, z, WATER);
                    }
                }
            }
        }
    }

    public Biome[] populateBiomes(Biome[] biomeBuffer, int chunkX, int chunkZ) {
        return this.world.getBiomeProvider().getBiomes(biomeBuffer, chunkX << 4, chunkZ << 4, 16, 16);
    }

    private void generateBiome(ChunkPrimer primer, int chunkX, int chunkZ) {
        double scale = 0.0625;

        int globalX = chunkX << 4;
        int globalZ = chunkZ << 4;

        this.depthBuffer = this.coverNoise.getRegion(this.depthBuffer, globalX, globalZ, 16, 16, scale, scale, 1.0);

        this.coverProvider.populate(this.regionHandler.get(), this.world, globalX, globalZ);
        CoverChunkDataProvider.Data coverData = this.coverProvider.getResultStore();
        Set<CoverType> coverTypes = coverData.getTypes();

        for (CoverType type : coverTypes) {
            CoverGenerator generator = this.generators.get(type);
            if (generator != null) {
                generator.getCover(this.random, globalX, globalZ);
                generator.getFiller(this.random, globalX, globalZ);
            }
        }

        super.coverBiome(primer, this.coverBlockBuffer, this.fillerBlockBuffer, this.depthBuffer, globalX, globalZ);

        if (this.settings.decorate && !this.fastGenerate) {
            GlobPrimer globPrimer = new GlobChunkPrimer(primer);
            for (CoverType type : coverTypes) {
                CoverGenerator generator = this.generators.get(type);
                if (generator != null) {
                    generator.coverDecorate(globPrimer, this.random, globalX, globalZ);
                }
            }
        }
    }

    @Override
    public void populate(int chunkX, int chunkZ) {
        BlockFalling.fallInstantly = true;

        int globalX = chunkX << 4;
        int globalZ = chunkZ << 4;

        this.coverProvider.populate(this.regionHandler.get(), this.world, globalX + 8, globalZ + 8);

        CoverChunkDataProvider.Data coverData = this.coverProvider.getResultStore();

        LatitudinalZone zone = this.getLatitudinalZone(globalX, globalZ);

        CoverType cover = coverData.getCoverData()[136];
        Biome biome = cover.getBiome(zone);

        if (this.settings.decorate && !this.fastGenerate) {
            ForgeEventFactory.onChunkPopulate(true, this, this.world, this.random, chunkX, chunkZ, false);

            Set<CoverType> coverTypes = coverData.getTypes();
            for (CoverType type : coverTypes) {
                CoverGenerator generator = this.generators.get(type);
                if (generator != null) {
                    generator.decorate(this.random, zone, globalX + 8, globalZ + 8);
                }
            }

            if (TerrainGen.populate(this, this.world, this.random, chunkX, chunkZ, false, PopulateChunkEvent.Populate.EventType.ANIMALS)) {
                WorldEntitySpawner.performWorldGenSpawning(this.world, biome, globalX + 8, globalZ + 8, 16, 16, this.random);
            }

            ForgeEventFactory.onChunkPopulate(false, this, this.world, this.random, chunkX, chunkZ, false);
        }

        BlockFalling.fallInstantly = false;
    }

    private LatitudinalZone getLatitudinalZone(int x, int z) {
        this.zoneScatterMap.initPosSeed(x, z);
        int offset = this.zoneScatterMap.nextInt(128) - this.zoneScatterMap.nextInt(128);
        Coordinate chunkCoordinate = Coordinate.fromBlock(this.settings, x + 8, z + 8 + offset);
        return LatitudinalZone.get(chunkCoordinate);
    }

    @Override
    public boolean generateStructures(Chunk chunk, int x, int z) {
        return false;
    }

    @Override
    public List<Biome.SpawnListEntry> getPossibleCreatures(EnumCreatureType creatureType, BlockPos pos) {
        return this.world.getBiome(pos).getSpawnableList(creatureType);
    }

    @Override
    public BlockPos getNearestStructurePos(World world, String structureName, BlockPos pos, boolean findUnexplored) {
        return pos;
    }

    @Override
    public void recreateStructures(Chunk chunk, int x, int z) {
    }

    @Override
    public boolean isInsideStructure(World world, String structureName, BlockPos pos) {
        return false;
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/world/generator/EarthChunkGenerator.kt
package net.gegy1000.terrarium.server.world.generator

import net.gegy1000.terrarium.server.capability.TerrariumCapabilities
import net.gegy1000.terrarium.server.map.glob.GlobGenerator
import net.gegy1000.terrarium.server.map.glob.GlobType
import net.gegy1000.terrarium.server.world.EarthGenerationSettings
import net.minecraft.block.BlockFalling
import net.minecraft.block.material.Material
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.EnumCreatureType
import net.minecraft.init.Blocks
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.WorldEntitySpawner
import net.minecraft.world.biome.Biome
import net.minecraft.world.chunk.Chunk
import net.minecraft.world.chunk.ChunkPrimer
import net.minecraft.world.gen.IChunkGenerator
import net.minecraft.world.gen.NoiseGeneratorPerlin
import net.minecraft.world.gen.layer.IntCache
import net.minecraftforge.event.ForgeEventFactory
import net.minecraftforge.event.terraingen.PopulateChunkEvent
import net.minecraftforge.event.terraingen.TerrainGen
import java.util.EnumMap
import java.util.Random
import kotlin.reflect.full.createInstance

class EarthChunkGenerator(val world: World, seed: Long, settingsString: String, private val fastGenerate: Boolean) : IChunkGenerator {
    companion object {
        private val STONE = Blocks.STONE.defaultState
        private val AIR = Blocks.AIR.defaultState
        private val BEDROCK = Blocks.BEDROCK.defaultState
        private val GRAVEL = Blocks.GRAVEL.defaultState
        private val RED_SANDSTONE = Blocks.RED_SANDSTONE.defaultState
        private val SANDSTONE = Blocks.SANDSTONE.defaultState
        private val ICE = Blocks.ICE.defaultState
        private val WATER = Blocks.WATER.defaultState

        private const val BEDROCK_HEIGHT = 5
    }

    val handler by lazy { world.getCapability(TerrariumCapabilities.worldDataCapability, null)!!.generationHandler }

    val random = Random(seed)
    val settings = EarthGenerationSettings.deserialize(settingsString)
    val coverNoise = NoiseGeneratorPerlin(this.random, 4)
    val pos = BlockPos.MutableBlockPos()

    val globGenerators = EnumMap<GlobType, GlobGenerator>(GlobType::class.java)

    var depthBuffer = DoubleArray(256)
    val heightBuffer = IntArray(256)
    val globBuffer = Array(256, { GlobType.NO_DATA })
    val coverBuffer = Array(256, { STONE })
    val fillerBuffer = Array(256, { STONE })

    var biomeBuffer: Array<Biome>? = null

    val maxHeight = this.world.height - 1

    init {
        GlobType.values().forEach {
            val generator = it.generator.createInstance()
            generator.initialize(this.world, this.globBuffer, this.heightBuffer, this.coverBuffer, this.fillerBuffer)
            this.globGenerators.put(it, generator)
        }
    }

    override fun generateChunk(chunkX: Int, chunkZ: Int): Chunk {
        val primer = generatePrimer(chunkX, chunkZ)
        this.biomeBuffer = this.world.biomeProvider.getBiomes(this.biomeBuffer, chunkX shl 4, chunkZ shl 4, 16, 16)

        val chunk = Chunk(this.world, primer, chunkX, chunkZ)

        biomeBuffer?.let { buffer ->
            val biomeArray = chunk.biomeArray
            for (biomeIndex in biomeArray.indices) {
                biomeArray[biomeIndex] = Biome.getIdForBiome(buffer[biomeIndex]).toByte()
            }
        }

        chunk.generateSkylightMap()

        return chunk
    }

    fun generatePrimer(chunkX: Int, chunkZ: Int): ChunkPrimer {
        this.random.setSeed(chunkX.toLong() * 341873128712L + chunkZ.toLong() * 132897987541L)

        val primer = ChunkPrimer()
        this.populateBlocks(primer, chunkX, chunkZ)
        this.generateBiome(primer, chunkX, chunkZ)
        return primer
    }

    private fun populateBlocks(primer: ChunkPrimer, chunkX: Int, chunkZ: Int) {
        val oceanHeight = this.handler.oceanHeight
        this.handler.populateHeightRegion(heightBuffer, chunkX, chunkZ)

        for (z in 0..15) {
            for (x in 0..15) {
                val height = heightBuffer[x + z * 16]
                for (y in 1..height) {
                    primer.setBlockState(x, y, z, STONE)
                }
                if (height < oceanHeight) {
                    for (y in height + 1..oceanHeight) {
                        primer.setBlockState(x, y, z, WATER)
                    }
                }
            }
        }

        if (this.settings.mapFeatures) {

        }
    }

    private fun generateBiome(primer: ChunkPrimer, chunkX: Int, chunkZ: Int) {
        val scale = 0.03125

        val globalX = chunkX shl 4
        val globalZ = chunkZ shl 4

        this.depthBuffer = this.coverNoise.getRegion(this.depthBuffer, globalX.toDouble(), globalZ.toDouble(), 16, 16, scale * 2.0, scale * 2.0, 1.0)

        this.handler.populateGlobRegion(globBuffer, chunkX, chunkZ)

        val generators = globBuffer.toHashSet()
        generators.forEach {
            this.globGenerators[it]?.let {
                it.getCover(globalX, globalZ, this.random)
                it.getFiller(globalX, globalZ, this.random)
            }
        }

        for (z in 0..15) {
            for (x in 0..15) {
                val index = x + z * 16
                val cover = coverBuffer[index]
                val filler = fillerBuffer[index]
                val noise = depthBuffer[index]
                this.generateBiomeTerrain(cover, filler, this.random, primer, globalX + x, globalZ + z, noise)
            }
        }

        if (this.settings.decorate && !fastGenerate) {
            generators.forEach {
                IntCache.resetIntCache()
                this.globGenerators[it]?.coverDecorate(primer, this.random, globalX, globalZ)
            }
        }
    }

    private fun generateBiomeTerrain(topBlock: IBlockState, fillerBlock: IBlockState, rand: Random, primer: ChunkPrimer, x: Int, z: Int, noise: Double) {
        val oceanHeight = this.handler.oceanHeight

        var currentTop = topBlock
        var currentFiller = fillerBlock
        var depth = -1
        val soilDepth = Math.max((noise / 3.0 + 3.0 + rand.nextDouble() * 0.25).toInt(), 1)
        val localX = x and 15
        val localZ = z and 15

        var localY = maxHeight
        while (localY >= 0) {
            if (localY < BEDROCK_HEIGHT) {
                if (localY == 0 || localY <= rand.nextInt(BEDROCK_HEIGHT)) {
                    primer.setBlockState(localX, localY, localZ, BEDROCK)
                }
            } else {
                var current = primer.getBlockState(localX, localY, localZ)
                while (current.material == Material.AIR && localY >= BEDROCK_HEIGHT) {
                    current = primer.getBlockState(localX, --localY, localZ)
                    depth = -1
                }
                if (current == STONE) {
                    if (depth == -1) {
                        if (soilDepth <= 0) {
                            currentTop = AIR
                            currentFiller = STONE
                        }
                        if (localY < oceanHeight && currentTop.material == Material.AIR) {
                            currentTop = WATER
                        }
                        depth = soilDepth

                        primer.setBlockState(localX, localY, localZ, currentTop)
                    } else if (depth-- > 0) {
                        primer.setBlockState(localX, localY, localZ, currentFiller)
                    } else {
                        localY = BEDROCK_HEIGHT
                    }
                }
            }
            localY--
        }
    }

    override fun populate(chunkX: Int, chunkZ: Int) {
        BlockFalling.fallInstantly = true

        this.handler.populateGlobRegion(globBuffer, chunkX, chunkZ)

        val x = chunkX shl 4
        val z = chunkZ shl 4

        this.random.setSeed(this.world.seed)
        val k = this.random.nextLong() / 2L * 2L + 1L
        val l = this.random.nextLong() / 2L * 2L + 1L
        this.random.setSeed(chunkX.toLong() * k + chunkZ.toLong() * l xor this.world.seed)

        val glob = globBuffer[255]
        val biome = glob.biome

        if (this.settings.decorate) {
            ForgeEventFactory.onChunkPopulate(true, this, this.world, this.random, chunkX, chunkZ, false)

            this.globGenerators[glob]?.decorate(this.random, x + 8, z + 8)

            if (TerrainGen.populate(this, this.world, this.random, chunkX, chunkZ, false, PopulateChunkEvent.Populate.EventType.ANIMALS)) {
                WorldEntitySpawner.performWorldGenSpawning(this.world, biome, x + 8, z + 8, 16, 16, this.random)
            }

            ForgeEventFactory.onChunkPopulate(false, this, this.world, this.random, chunkX, chunkZ, false)
        }

        BlockFalling.fallInstantly = false
    }

    override fun getPossibleCreatures(type: EnumCreatureType, pos: BlockPos): MutableList<Biome.SpawnListEntry> {
        return this.world.getBiome(pos).getSpawnableList(type)
    }

    override fun getNearestStructurePos(world: World, structureName: String, pos: BlockPos, map: Boolean) = pos

    override fun generateStructures(chunk: Chunk, x: Int, z: Int) = false

    override fun recreateStructures(chunk: Chunk, x: Int, z: Int) {
    }

    override fun isInsideStructure(worldIn: World?, structureName: String?, pos: BlockPos?): Boolean {
        return false
    }
}
